File: .github/scripts/generate-release-notes.py
# Credit: https://github.com/metafates/mangal
import pathlib as pl

IN = "CHANGELOG.md"
OUT = "whats-new.md"

# get script path
script_path = pl.Path(__file__).resolve()

# get project root (3 levels up)
project_root = script_path.parent.parent.parent

# get changelog path
changelog_path = pl.Path(project_root, IN)

# get changelog content
with open(changelog_path, "r") as f:
    changelog = f.read()

    # we need to extract everything between the first and the second header with tags should not remove other header tags
    changelog = changelog.split("## ")[1]  # remove everything before the first header
    changelog = changelog.split("## ")[0]  # remove everything after the second header

    # remove the first line
    changelog = "\n".join(changelog.split("\n")[1:])

    # trim newlines
    changelog = changelog.strip()

    # write to file
    with open(pl.Path(project_root, OUT), "w") as ft:
        ft.write(changelog)

File: .github/scripts/generate_release_notes.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

func main() {
	const inFile = "CHANGELOG.md"
	const outFile = "whats-new.md"

	// Get the path to the changelog
	changelogPath := filepath.Join(".", inFile)

	// Read the changelog content
	content, err := os.ReadFile(changelogPath)
	if err != nil {
		fmt.Println("Error reading changelog:", err)
		return
	}

	// Convert the content to a string
	changelog := string(content)

	// Extract everything between the first and second "## " headers
	sections := strings.Split(changelog, "## ")
	if len(sections) < 2 {
		fmt.Println("Not enough headers found in the changelog.")
		return
	}

	// We only care about the first section
	changelog = sections[1]

	// Remove everything after the next header (if any)
	changelog = strings.Split(changelog, "## ")[0]

	// Remove the first line (which is the title of the first section)
	lines := strings.Split(changelog, "\n")
	if len(lines) > 1 {
		changelog = strings.Join(lines[1:], "\n")
	}

	// Trim newlines
	changelog = strings.TrimSpace(changelog)

	// Write the extracted content to the output file
	outPath := filepath.Join(".", outFile)
	if err := os.WriteFile(outPath, []byte(changelog), 0644); err != nil {
		fmt.Println("Error writing to file:", err)
		return
	}

	fmt.Printf("Changelog content written to %s\n", outPath)
}
